### ISSUE TITLE
Plugin Package Manager

### ISSUE THREAD

**Issue by dvidsilva** [neovim/neovim#247]
Hi, I might speak for many when I say that package managers for the vim plugins are not like awesome.

is this already been considered, something like nvpm (neovim package manager)

would someone wanna join me/start  in this task

**Comment by addisonamiri** [comment]
Now that the lua plugin implementation is a little further along, are there any new developments on this subject?

TL;DR: There should be a standard repo structure, a default repo, and the ability to add 3rd party repos easily.

I also would like to add my support for a vim-pi-like package manager. I think a repository (git or otherwise) with a bunch of build files in it would be the best option for a package manager. I think that build files close to ArchLinux pkgbuilds it would be a decent solution though others may have better ideas. The fields I'm thinking of off the top of my head would be dependencies, a source for the plugin (git or otherwise), build requirements (python/ruby/lua support), a version number, conflicts, provides, supported releases of neovim and so on.

While the build file and repository structure should be standard, the repository location should not be. If at any point github goes down, or any repo becomes stale, someone could fork the repo to another source and people could add the new repo to their neovim config.

For a default repo I think an anarchic one would be best as to not burden the developers of Neovim into package maintinence as well as give plugin developers an easy way to distribute their plugin. What I'm thinking of is this. To submit a package the plugin developer could make a pull request to the default repo adding their build file. A hook in the repo could check for syntactic correctness and accept the pull. The developer of the plugin would then be given exclusive write access to that file and nothing else (to push updates). I can't think of anything for access changes but if a plugin developer is handing control to someone else (though rarely the case I think it should be encouraged over forks) it should be supported.

A consideration should be taken to not rely on any specific version control system. This would allow users to create their own local repo and build files to test their own plugins and add those not in the remote repo. Also while there should be a default repo I'd like to stress there is nothing stopping someone else from creating a new public repo with different submission guidelines (if someone in the community can spend more time curating the list than neovim developers can ala vimawesome). Also this solves the problem previously discussed of moving from one source to another. Just bump the version number and change the source.

Implementation aside, I think the important things to consider is having a default that's easily accessible to plugin developers (even at the cost of quality) and being as open to third party repos as possible.

Although it's one of the ancient unix wars, I think emacs' plugin infrastructure can be a good reference on the subject. From the experience of a new user, it seems to be working pretty well. Basically all I had to do was add a third party repo and I was good for everything I wanted. Something like this is very easy to add in a getting started guide (over the whole step 1: pick a package manager in current vim).

**Comment by Shougo** [comment]
Hi, I'm neobundle author.

Yes, I think neovim or Vim should provide standard repository architecture.
It is important what is the standard plugin manager.

**Comment by tim-lebedkov** [comment]
The most important first question in my opinion about a package manager for Neovim is about its type.

As far as I can see from the comments above the following **types** were proposed:
* use native package manager (Apt for Linux, Homebrew for MacOS, etc.)
* Vimscript based package managers with an API provided by Neovim
* one built-in Neovim specific package manager (e.g. built in Lua)
* use a cross-platform, not Neovim-specific package manager (Bower, etc.)

How can we answer this question? Maybe a **poll** would give at least some hints about what should be done?

**Comment by fmoralesc** [comment]
People, [gmarik from Vundle is asking for maintainers](https://github.com/gmarik/Vundle.vim/issues/608). Do you think it would be much of a long shot to take over that project under the neovim org? There is certainly interest in including a plugin manager by default in neovim, and there's been much discussion on what it should do. Having Vundle under neovim's umbrella might make it easier to move forward in this front.

**Comment by justinmk** [comment]
> Do you think it would be much of a long shot to take over that project under the neovim org?

I'd rather not. vim-plug is better designed, less baggage. In any case it will be a quagmire that distracts us from more pressing goals.

**Comment by fmoralesc** [comment]
I agree that vim-plug is cleaner, and this all distracts from other goals, but I just wanted to point it out in case someone was interested.

**Comment by hauleth** [comment]
@tim-lebedkov:

1. This doesn't work. Official repositories are slow (to update) so this one is dead end (also no one is doing that AFAIK)
2. IMHO this one isn't much different from 3.
3. IMHO the best option, as we can build centralized repository for packages like Hex/RubyGems/NPM  and maintain it in easy way.
4. Silly idea as it is polluting some tool niche with something completely unrelated. This would only piss people off.

There should be no discussion in this matter. Someone in core team should make this and close this pointless discussion. Everyone need that but if we let people discuss that matter then there will never be consensus. Someone must make it "happen" - just like divine intervention.

**Comment by justinmk** [comment]
> Everyone need that but if we let people discuss that matter then there will never be consensus.

We're very aware of that and we don't use polling to make decisions. But the current state of the plugin manager discussion is that ideas are welcome because we don't have the resources to make something happen with it right now. Putting up a centralized repo is a permanent commitment, not something to flirt with.

**Comment by ZyX-I** [comment]
@hauleth 

1. Some package managers are able to install “development versions”: e.g. the so-called live ebuilds in Gentoo portage. Mainly applicable for source distributions. Official repositories are not a problem: I do not know package managers which do not allow you to maintain your own repository. In fact ability to work with users own repository is a *requirement* for any normal system package manager because this is how software is distributed *locally in companies* when it needs to be installed on a multiple similar machines.

    Problem with this point is not slow reaction or something like this. Problem here is that we will need to support *multiple* package managers here.
2. Do not really understand how it should actually look like.
3. I also think this is the best option.
4. Why polluting? Some of such package managers also allow custom repositories. This would be second-best option if I knew any generic enough package managers. All package managers I know fall into one of the following categories:

    1. Distribution-specific “system” package managers like portage (Gentoo) or apt (Debian/-based). These are inappropriate because of being too tied up with distribution. Most of them even cannot install into $HOME.
    2. Language-specific package managers like pip (Python). Inappropriate because they mostly have no “other languages” support and (almost: compiled languages excluded) always require having compiler/interpreter of the language they are written for: this way only luarocks may be used (because lua is a requirement).
    3. Scope-specific package managers like bower (web development). Inappropriate because if scope is not “managing plugins for the editor” you will get different inconveniences just because developers were not preparing their software for this use-case, exact inconveniences depend on the package manager scope.

    Maybe Nix package manager can handle our needs, probably with some additional support from NeoVim: at least, it is one of the rare system package managers that allows installation to user directory.

The only “proper” centralized repository I know is [new vim-pi](https://bitbucket.org/vimcommunity/vim-pi-db) ([documentation](https://bitbucket.org/vimcommunity/vim-pi-documentation)), but it is not finished. Existing and used vim-pi database is rather inconvenient to use and has many flaws: single version support (also meaning that you cannot specify version of the dependency), VimL language for hooks (hooks are used to make “build scripts”), inconsistent keys in addon-info.json.

**Comment by fmoralesc** [comment]
I just went ahead and made some changes to include vim-plug into neovim's runtime, which you can test from https://github.com/fmoralesc/neovim/tree/early/pluginmanager

So, big deal, that's super lame, right? Well, that branch includes some other changes that make it possible for the user not having to deal with the `call plug#begin()` and `call plug#end()` stuff, so the `:Plug*` commands are always available. I think that is a win from the user perspective. I have to test how that behaves alongside my #2675 PR.

I know this is less interesting than working on the repository stuff, but it's quite a lot more than nothing. We can't expect to figure out all the issues or having someone to come up with a perfectly formed solution, so I believe we should instead focus in incrementally improving the user experience.

I made some slight changes to the vim-plug plugin, but only in the comments. The documentation of the plugin should be extracted to runtime/doc.

**Comment by justinmk** [comment]
> some other changes that make it possible for the user not having to deal with the call plug#begin() and call plug#end() stuff,

That's a good point. However I have concern about #2675 which I will mention there.

**Comment by weaksauce** [comment]
does this prevent someone from using, say, vundle or some other plugin manager? if not it seems like a no brainer to include it. (I am a little biased because vim-plug is very well done and the one that I use after trying many of the available options)

Either way it wouldn't change my life that much though.

**Comment by fmoralesc** [comment]
> does this prevent someone from using, say, vundle or some other plugin manager? 

No, as long as there isn't a clash with the names of the commands. I would see very little reason to use anything else than vim-plug if included, though (I am also biased).

**Comment by fmoralesc** [comment]
> Oops I thought that was the "early" branch.

So, are there concerns about the early plugins stuff, or was it about the mapping defaults (which I removed from the PR)?

**Comment by tim-lebedkov** [comment]
I've created a **poll** for the plugin manager **type**. As I already wrote, it should only be a decision help. BTW, option number 2 is for Pathogen, Vundle and similar plugin managers.

You can vote here:
https://docs.google.com/forms/d/1finx03Wddt6Uw4TwVXLARD4Dd9pXp9dg4295U9itldQ/viewform?usp=send_form

Results: https://docs.google.com/forms/d/1finx03Wddt6Uw4TwVXLARD4Dd9pXp9dg4295U9itldQ/viewanalytics

**Comment by tim-lebedkov** [comment]
It's a pity so few people chose a native package manager (option 1). Here is an example of how this could work: https://www.youtube.com/watch?v=xbcjvO_HvYM

**Comment by ZyX-I** [comment]
@tim-lebedkov *Regardless* of the result of *our* choice native package manager *will* be used. But native plugin managers are naturally limited to plugins that are interesting to at least one of the repository (or overlay/ppa/…) maintainers and thus contain only small amount of plugins.

Once plugin manager and plugin repository is chosen it will be possible to create scripts like g-cpan ([g-cpan: generate and install CPAN modules using portage](http://www.gentoo.org/proj/en/perl/g-cpan.xml)) which can be used by maintainers or users for generating packages or build-install scripts (e.g. ebuilds) automatically.

---

Also note that without choosing domain-specific plugin description and/or repository formats it would be impossible to use *any* option. By the way, new [plugin-info.json specification](https://vim-plugin-index.readthedocs.org/en/latest/) and new [vim-pi](https://bitbucket.org/vimcommunity/vim-pi-db) repository is way closer to being complete: everything which can be generated automatically from www.vim.org was finished (except that existing plugins’ archives and SCM repositories are not searched neither for addon-info.json/`{name}-plugin-info.txt`/`plugin-info.json` files nor for GLVS-style dependency descriptions; old vim-pi cron job update script neither does this), hooks and old SCM sources from current vim-pi were not converted, but can already be.

**Comment by zaxebo1** [comment]
@ZyX-I : **If you want to use native package manager** , then we need to look into these issues:

1) how will you handle this on Vim(neovim) **for MS-Windows** . Many of us are using vim/neovim on windows heavily.

2) On slackware Linux, how are you going to handle this, for which there is no online repository inbuilt package manger (like yum,apt-get,..). Though users have contributed third party extensions for slackware, but there are many such third party optional native repository tool for slackware which is inbuilt.

**Comment by ZyX-I** [comment]
@zaxebo1 I am not saying that native package managers are good choice. In fact I’ve just provided yet another reason against relying on them. And a few messages earlier I said that NeoVim-specific plugin manager is the best option.

By the way, there is another option for 4. (non-NeoVim-specific non-system package manager): linuxbrew. At least, unlike nix, I was able to install and successfully use it. But this is again non-Windows, at least currently.

**Comment by saep** [comment]
I think that users of vanilla neovim should be able to handle the complexity of installing vim-plug, Vundle, Pathogen etc.

A specific (GUI) distribution of neovim could offer a built-in plugin manager to make discovery and installation pleasant. But I think it is best left to third parties. (A possible example for this could be that eclipse decides to use neovim as its new editor backend and creates a plugin manager that integrates with the current infrastrurcture (market-place, update-sites).)

Some plugin managers define special syntax for scripts hosted on github. Although this is reasonable to do at the moment, a plugin manager that ships with neovim should not default to special services, especially if they are tied to a domain.

It is exceptionally hard to create a plugin manager that can generically handle arbitrary languages for plugins. This means, that it should work on any operating system, that you do not need administrator priviliges to use it and that it is more convenient to use as current plugin managers.

In any case, if neovim decides to ship with some form of plugin manager, it is probably a good idea to take a look at emacs. I don't think it would hurt to define (or decide to use) some kind of metadata format (e.g. vim-pi) and create a GUI around it, but the service to access them must be configurable (and not bound to a specific domain) and the neovim project should manage one repository on its own. The repository must not be big, but it should work as some kind of reference implementation. It could offer a place for language-specific plugins that essentially contain enough vimL (or lua) to install other plugin providers. This would mean, that a neovim environment could be bootstrapped by a single .nvimrc file. (Well, it could be done already, but it would require more work.)

**Comment by ZyX-I** [comment]
New Vim-PI repository format definition says nothing about specific service. In fact I was going to create a file which will list all “raw” URLs that can be used to access the repository (I think it is way too big to use `git clone` under most circumstances, even `git clone --depth=1`) and assume (and document) that plugin managers should use this file to determine how to check out the repository: this makes it possible to

1. Move all indexes and manifests to a separate repository (not going to actually do this now, but this move was being considered): these are purely-generated files (non-purely-generated: automatic plugin descriptions based on www.vim.org: they can be extended by users and also github-only plugins will be written and not generated).
2. Move all repositories from a bitbucket and still not require using `git clone`. (Note: if the chosen format is not a DSL (it will not be), services to move to will be limited to those that have rather simple raw check outs (e.g. no `foo.org/vimcommunity/vim-pi-db/raw?file=_i_n_d_e_x_._j_s_o_n`).)

**Comment by dfroger** [comment]
What about using then [miniconda](http://conda.pydata.org/miniconda.html) package manager?

It's cross-plateform and can install all kind of dependencies (from example Python). I would make it trival for a Vim plugin to depend for example on a C compiled (or other language) dependency (for example, a linter, a debogger....).

Another good point is that user could use conda environments.

**Comment by addisonamiri** [comment]
What is the decision process for what ends up getting implemented? Everyone seems to have given good arguments. I'd suggest a community poll but from what I saw on the default colorscheme there didn't seem to be too many votes. The developers of Neovim and current package managers seem to be the most knowledgeable so even if it's just them having a discussion and choosing one that'd be fine. I would just like to see one of the options chosen so it can become _actionable_.

**Comment by weaksauce** [comment]
Every decision to add something comes at a cost of later support and dependencies. Something like a package manager in neovim would have a lasting impact so it's wise to defer final judgement onto someone who knows the software well and has a thorough understanding of the most promising options. I don't think a poll is a good option because it's limited in discussion about the pros and cons of different approaches.

**Comment by MarcWeber** [comment]
@dvidsilva You made a statement: All plugin managers for Vim suck. Eventually tell why - you may surprised that VAM and Neobundle are likely to be just "good enough". Yes there are still problems. But finding out which plugins to use and reading their documentation is more work IMHO. VAM is intergrated into nixos.org linux. You can create a .vim-addons file having a json dict per line and make vim create .nix files - thus have your linux distro install all plugins - eg I get a vim-marc-weber in PATH without disturbing others.

What are the problems? A lot of people don't want to care or learn about 10 different distros or ways to install plugins. Often plugins change more often than distros. Thus native plugin managers can work - but they may or may not be best. Also some plugins require "vim with python" or vim with ruby" or "vim with python version XYZ and python plugins ABC"... for this reason I had the idea of creating a plugin manager ruling all languages/tools to get rid of the hell, a draft / sketchy description is here: https://github.com/code-once/ypm
Eg if you install ruby/haskell/python/.. packages you often need headers of mysql or whatnot.

A somewhat comprehensive list of plugin managers (probably little bit outdated):
http://vim-wiki.mawercer.de/wiki/topic/vim%20plugin%20managment.html

Please also note that its important to not only have all plugins available, but also to give the guidance about which plugins "just work" - eg there are 5+ plugins which do syntax checking - some do only syntax checking (vim-addon-syntax-checker) - some mix syntax checking and running interpreters (syntastic) - which is why I forked it once. For this reason if people often report problems VAM allows to deprecate plugins giving a reason why and which pluigns might be tried instead.

If you're serious about security you should /must review the code of all plugins .. or build a trust system as well. For this reason setting up something the linux kernel has (maintainers and trusting them) could be nice.

Thus getting the plugin story right is little bit more than fetching stuff from github (which you also want to be able to do - and which all major plugin managers allow doing).

Vundle does implement a search - but was limited to "github" AFAIK. VAM implements a search, but its limitd to plugin names. http://vam.mawercer.de/ is a list of plugins VAM knows about.

Things like "using gems" has been tried before - but unless the majority supports it it is likely to fail.
As VAM author we also noticed how important it is to activate plugins "as needed" for startup speed reasons. You can even install plugins from within VIM without restarting.

I had the "one-vim-lib" attempt - allowing the user to configure mappings. I had code turning a vim dictionary into a tree so that you can edit it... I even spent huge amounts of time on allowing to activate/deactivate parts - just to learn than viml is too slow - and that removing mappings was more complicated than restarting Vim. Since then I try to keep things simple (I still fail sometimes).

About mappings: I find its a nice compromise to have default mappings but allow the user to opt-out and define his/her own. In the snipmate case there is a dictionary allowing you to override the mappings. For plugins which are likely to collide with others I find this is a nice option:
Having a s:opts buffer local dicitonary which gets bound to a g:plugin_name as well.
Then users ca preset options. This worked quite well for almost all plugins I worked on.
http://vim-wiki.mawercer.de/wiki/topic/designing-a-vim-plugin.html => configuration

**Comment by sullyj3** [comment]
- I vote in favour of autocratic decision making :)
- Having said that, I really like vim-plug as well. I'd like to see it or something similar shipped by default and maintained by neovim.

> Some plugin managers define special syntax for scripts hosted on github. Although this is reasonable to do at the moment, a plugin manager that ships with neovim should not default to special services, especially if they are tied to a domain.

I kind of agree. In vim-plug, something like "Plug 'tpope/vim-surround'" grabs from https://github.com/tpope/vim-surround. I think it would be best if the default plugin manager shipped with a simple functionality. When it encounters a Plugin string that seems to be only part of a URL, look through a configurable list of url prefixes in order and install the first 'tpope/vim-surround' it finds. The default nvimrc could ship with a list of default URL prefixes to be searched, beginning with "https://github.com/". This list could be updated as different hosting platforms waxed and waned in popularity (raising the obvious problem of how to decide which ones to include). 

Another alternative could be to include and document the functionality for configuring URL prefixes to search, but not including any list of defaults. IMO, this is the best alternative.

**Comment by ZzAntares** [comment]
my two cents...

- Theres no need to have a native package manager, instead a plugin should do it, and when plugins become so awesome then are elegible to be natively included in the neovim distribution.

- Please **do not** create a centralized plugin repository, no npm or rubygems like please, is better the [packagist.org](https://packagist.org/) approach where packagist is only an index of submitted libraries (as @tarruda says), with an account that can be associated with github, notice the vendor feature for better package organization, it only fetches CVS, it could work with git/mercurial and integrated for searching in github, bitbucket or even private repos. Such index it could even have an awesome API that could allow navigate the index within nvim, even read doc files, 

- Define a standard for plugin packaging via a universal agnostic programming language spec file format, like `json`, `yaml`, `xml`, etc, something clean and readable.

- Encourage to use semantic versioning for plugin packaging and this will allow the nvim package manager to work with version constraints, allowing an ecosystem where plugins could define dependencies on other plugins (imagine a BDD plugin for testing plugins, or a rake like plugin), also will allow avoid conflicts between them.

- Multi language support  for the package manager is not needed just yet!, firs an extensible package manager is needed, one that even could have plugins hooked into it, and then such plugins could easily integrate the compatibility of requiring plugins written in other programming languages.

But the most important thing! for now just define and document the plugin architecture that nvim will use, because at the moment changes on the new architecture stops anyone to start building this awesome plugins. I also agree with @aktau 

> The ability for other plugins systems over msgpack-rpc is just icing on the cake.

**Comment by hauleth** [comment]
@ZzAntares what is wrong with centralized package repository? Fetching packages from GitHub has one big disadvantage:

> Even if one use semantic versioning, one could still f… it up.

I mean, Git tags aren't immutable. You always can delete it and create new, then force push. This is horrible and it is main reason why most of FE packages migrate from Bower to NPM - NPM doesn't allow you to change once submitted version.

**Comment by ZzAntares** [comment]
@hauleth I think is a better approach, that way is just an index not a storage service, and in addition no one has to worry about storage costs on the hosting server, mirroring is easier since source is always in a repo and there's no need to sync packages but only the index. Even if a tag is removed that doesn't change nothing when requiring a dependency, take a look at composer/packagist which uses versions constraints that ensure than even after deleting a tag, a dependency is still resolvable not to that tag but to the best match, one can even reference branches or plain commits, also they use a neat setting called the _minimum stability_ worth to watch.

Besides that vim/nvim community is well versed in cvs (this is not sublime community, we're sick people haha), we know a tag must not be deleted, but as I told you, even if it goes deleted is not a problem, version constraints are the solution to that.

Just my opinion though.

**Comment by Shougo** [comment]
https://glyph.twistedmatrix.com/2015/11/editor-malware.html

vim.org does not support https.  I think it is fatal problem.
We should provide another secure repository for the plugins.

**Comment by josephholsten** [comment]
@Shougo I totally agree, but this is almost certainly not the right ticket for that discussion. Also, it probably would be good to get the folks interested in http://vim-scripts.org and http://vimawesome.com involved. If we can get them both to stop encouraging insecure plugin sources, it'll be one thing that isn't on our plate.

**Comment by Shougo** [comment]
> @Shougo I totally agree, but this is almost certainly not the right ticket for that discussion. Also, it probably would be good to get the folks interested in http://vim-scripts.org and http://vimawesome.com involved. If we can get them both to stop encouraging insecure plugin sources, it'll be one thing that isn't on our plate.

OK.  The plugins repository issue is already available?
If not, I will create it.

I think vimawesome is better place.  But it is not updated until March.
The author seems be busy.  I think another maintainer is needed.

**Comment by jasper-lyons** [comment]
my 2c

I would suggest applying a little product development to this? Why don't we figure out what the ideal end situation is and then try and identify a rough, iterative path between there we are now. e.g.

skate board -> scooter -> unicycle -> bicycle -> moped -> motorbike.

Then at each stage we can see if people actually give a damn before committing to the next and wasting a whole bunch of really smart peoples time. Instead of deciding it's a great idea, spending 18 months building the whole thing and then finding out vim-plug just does a better job.

**Comment by justinmk** [comment]
Very releavant article by an expert: [So you want to write a package manager](https://medium.com/@sdboyer/so-you-want-to-write-a-package-manager-4ae9c17d9527)

**Comment by justinmk** [comment]
Continued in https://github.com/neovim/neovim/issues/20893
